pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,308 @@
+		BeforeEach(func() {
+			corev1.AddToScheme(scheme)
+			tmpFederator := newFederatorWithScheme(scheme)
+			fedResource, _ = tmpFederator.createFederatedResource(resourceTyped)
+		})
+
+		JustBeforeEach(func() {
+			fedPod = getResourceFromAPI(f.kubeFedClient, fedResource)
+		})
+
+		It("should return no error on success", func() {
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		When("the resource is already in the Kube API", func() {
+			expectedImage := "apache"
+
+			BeforeEach(func() {
+				initObjs = append(initObjs, fedResource)
+
+				resourceTyped.Spec = corev1.PodSpec{
+					Containers: []corev1.Container{
+						corev1.Container{
+							Image: expectedImage,
+							Name:  "httpd",
+						},
+					},
+				}
+			})
+
+			It("should update the resource", func() {
+				fedPodObj, _, _ := getTemplateField(fedPod)
+				expectObj, _, _ := getTemplateField(fedResource)
+
+				Expect(fedPodObj["spec"]).ToNot(BeEquivalentTo(expectObj["spec"]))
+
+				mapObj, _, _ := unstructured.NestedSlice(
+					fedPod.Object,
+					ctlutil.SpecField, ctlutil.TemplateField, "spec", "containers",
+				)
+				image := mapObj[0].(map[string]interface{})["image"]
+				Expect(image).To(BeEquivalentTo(expectedImage))
+			})
+		})
+
+		When("the resource is not in the Kube API", func() {
+			It("should create the resource", func() {
+				Expect(fedPod).ToNot(BeNil())
+				Expect(fedPod).To(BeEquivalentTo(fedResource))
+			})
+		})
+
+		When("the resource needs to be distributed to specific clusters", func() {
+			BeforeEach(func() {
+				clusterNames = []string{"cluster1", "cluster2"}
+			})
+
+			It("should set the placement to be an explicit list of clusters", func() {
+				mapObj, found, err := getClustersField(fedPod)
+				fedClusterNames := unpackClusterNames(mapObj)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
+				Expect(fedClusterNames).To(BeEquivalentTo(clusterNames))
+			})
+
+			It("the returned resource's placement should be an empty matcher", func() {
+				// This test implicitly tests that the nested hierarchy of fields exists
+				mapObj, found, err := getMatchLabelsField(fedPod)
+				Expect(err).ToNot(HaveOccurred())
+				Expect(found).ToNot(BeTrue(), "Hierarchy of fields is present!")
+				Expect(mapObj).To(BeEmpty())
+			})
+		})
+	})
+}
+
+func newPod(name string) *corev1.Pod {
+	resourceTyped := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{Kind: "Pod"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: "test-namespace",
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				corev1.Container{
+					Image: "nginx",
+					Name:  "httpd",
+				},
+			},
+		},
+	}
+	return resourceTyped
+}
+
+func newFederatorWithScheme(scheme *runtime.Scheme, initObjs ...runtime.Object) *federator {
+	f := &federator{
+		scheme: scheme,
+		kubeFedClient: fake.NewFakeClientWithScheme(
+			scheme,
+			initObjs...,
+		),
+	}
+	return f
```

Comment:
nit: Why not simply do the return of the instance? Why do you need a named variable?